manticore/core/smtlib/solver.py

```diff
@@ -658,22 +636,142 @@ def get_value(self, constraints: ConstraintSet, *expressions):
+            "(set-option :tactic.solve_eqs.context_solve false)",
+        ]
+        command = f"{consts.z3_bin} -t:{consts.timeout * 1000} -memory:{consts.memory} -smt2 -in"
+
+        support_minmax, support_reset = self.__autoconfig()
+        super().__init__(
+            command=command,
+            init=init,
+            value_fmt=16,
+            support_minmax=False,
+            support_reset=True,
+            support_pushpop=True,
+            debug=False,
+        )
+
+    def __autoconfig(self):
+        # To cache what get-info returned; can be directly set when writing tests
+        self.version = self._solver_version()
+        if self.version >= Version(4, 5, 0):
+            support_minmax = False
+            support_reset = False
+        elif self.version >= Version(4, 4, 1):
+            support_minmax = True
+            support_reset = False
+        else:
+            logger.debug(" Please install Z3 4.4.1 or newer to get optimization support")
+        return support_minmax, support_reset
+
+    def _solver_version(self) -> Version:
+        """
+        If we fail to parse the version, we assume z3's output has changed, meaning it's a newer
+        version than what's used now, and therefore ok.
+
+        Anticipated version_cmd_output format: 'Z3 version 4.4.2'
+                                               'Z3 version 4.4.5 - 64 bit - build hashcode $Z3GITHASH'
+        """
+        try:
+            received_version = check_output([f"{consts.z3_bin}", "--version"])
+            Z3VERSION = re.compile(
+                r".*(?P<major>([0-9]+))\.(?P<minor>([0-9]+))\.(?P<patch>([0-9]+)).*"
+            )
+            m = Z3VERSION.match(received_version.decode("utf-8"))
+            major, minor, patch = map(
+                int, (m.group("major"), m.group("minor"), m.group("patch"))  # type: ignore
+            )
+            parsed_version = Version(major, minor, patch)
+        except (ValueError, TypeError) as e:
+            logger.warning(
+                f"Could not parse Z3 version: '{str(received_version)}'. Assuming compatibility."
+            )
+            parsed_version = Version(float("inf"), float("inf"), float("inf"))
+        return parsed_version
+
+
+class YicesSolver(SMTLIBSolver):
+    def __init__(self):
+        init = ["(set-logic QF_AUFBV)"]
+        command = f"yices-smt2 --timeout={consts.timeout * 1000}  --incremental"
+        super().__init__(
+            command=command,
+            init=init,
+            value_fmt=2,
+            debug=False,
+            support_minmax=False,
+            support_reset=False,
+        )
+
+
+class CVC4Solver(SMTLIBSolver):
+    def __init__(self):
+        init = ["(set-logic QF_AUFBV)", "(set-option :produce-models true)"]
+        command = f"cvc4 --lang=smt2 --incremental"
+        super().__init__(command=command, value_fmt=10, init=init)
+
+
+class ddRaceSolver(SMTLIBSolver):
+    def __init__(self, *solvers):
```

Comment:
I think this class deserves a brief explanation about what is going to happen

Reply:
HAHA. Yes.